datastore/adapter/keytransform.py

```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
+	_default_sharding_key: datastore.Key = datastore.Key("SHARDING")
+	_default_sharding_func: str = "v1/prefix/2"
+
+	@classmethod
+	async def create(
+			cls: typing.Type[T],
+			child_datastore: DS,
+			*args,
+			sharding_key: datastore.Key = None,
+			default_sharding_func: str = None,
+			**kwargs
+	) -> T:
+		"""Initializes _FlatFSAdapter with sharding function.
+
+		Arguments
+		---------
+		cls
+			Class variable passed in due to `classmethod`
+		child_datastore
+			Underlying datastore always used to initialize an `Adapter`
+		*args
+			Variable length list argument to be passed to cls constructor
+		sharding_key
+			Key used to access the sharding function in the underlying Datastore
+		default_sharding_func
+			Default sharding func to use if one is not found in the datastore
+		"""
+		# Get the variables needed to parse the sharding function
+		sharding_key = (
+			sharding_key
+			if sharding_key is not None
+			else cls._default_sharding_key  # type: ignore[attr-defined] # noqa: F821
+		)
+		default_sharding_func = (
+			default_sharding_func
+			if default_sharding_func is not None
+			else f"{cls.prefix}{cls._default_sharding_func}"   # type: ignore[attr-defined] # noqa: F821
+		)
+		sharding_fn: KEY_TRANSFORM_T = await \
+			cls._parse_sharding_function(  # type: ignore[attr-defined] # noqa: F821
+				child_datastore, sharding_key, default_sharding_func
+			)
+
+		# Return an instantiated version of the class
+		return cls(
+			child_datastore, *args, key_transform=sharding_fn, **kwargs
+			)  # type: ignore[attr-defined] # noqa: F821
+
+	@staticmethod
+	def _prefix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).ljust(length, "_")[:length])
+	
+	@staticmethod
+	def _suffix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length, "_")[:length])
+
+	@staticmethod
+	def _next_to_last(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length + 1, "_")[:length])
+
+	@classmethod
+	async def _parse_sharding_function(
+			cls: typing.Type[T],
+			child_datastore: DS,
+			sharding_key: datastore.Key,
+			default_sharding_func: str
+	) -> KEY_TRANSFORM_T:
+		"""Determine the sharding function from the sharding_key.
+		
+		Arguments
+		---------
+		cls
+			Class variable passed in due to `classmethod`
+		child_datastore
+			Underlying datastore always used to access the `sharding_key`
+		sharding_key
+			Key used to access the sharding function in the `child_datastore`
+		default_sharding_func
+			Default sharding func to use if one is not found in `child_datastore`
+
+		Raises
+		------
+		Exception
+			Empty shard identifier file.
+		Exception
+			Prefix was not present in sharding file
+		Exception
+			Sharding file was in incorrect format
+		Exception
+			Expect `v1` format
+		Exception
+			`key_length` shouold be an integer
+		Exception
+			`function_name` was not `prefix`, `suffix`, or `next-to-last`
```

Comment:
Doing `raise Exception("<some text>")` is an anti-pattern in Python (unlike JavaScript where the exception type is mostly useless anyways). Please add an exception hierarchy like the following for this:

```py
class InitializationError(RuntimeError):
	"""Raised when there was an error with opening this datastore adapter."""
	datastore: T
	
	def __init__(self, datastore: T, *args):
		self.datastore = datastore
		super().__init__(*args)

# Use for other errors
class ShardingStateError(InitializationError):
	"""Raised when the existing sharding entry could not be read."""
	key: datastore.Key
	
	def __init__(self, datastore: T, key: datastore.Key, *args):
		self.key = key
		super().__init__(datastore, *args)

# Use for: Empty shard identifier file.
class ShardingStateEmptyError(ShardingStateError):
	"""Raised when the existing sharding entry is empty."""

# Use for: Expect `v1` format & `function_name` was not `prefix`, `suffix`, or `next-to-last`
class ShardingStateUnsupportedError(ShardingStateError):
	"""Raised when the existing sharding entry uses an unsupported version number or sharding function."""
```

(The comment lines are just for your orientation.) Add this to the file top, add the relevant entries to `__all__` and update all raise statements to `raise <ErrorType>(self, sharding_key, "<some text>")`. This way consumers of this API can more easily distinguish between the kinds of errors raised (using `try: …; except <ErrorType> as exc: …`). Also some extra error data is included to help callers understand in which context the error occurred.